File: crates/kuben-platform/src/registry.rs
//! Cluster registry. Immutable after construction; every operation names its
//! cluster explicitly (Invariant I-11). Phase 0 supports a single cluster;
//! the type is already keyed so multi-cluster (phase 3) is additive.

use std::{collections::BTreeMap, fmt, sync::Arc};

use kube::{Client, Config, config::KubeConfigOptions};
use kuben_core::config::KubeCfg;

#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct ClusterId(Arc<str>);

impl ClusterId {
    pub const PRIMARY: &'static str = "primary";

    #[must_use]
    pub fn primary() -> Self {
        Self(Arc::from(Self::PRIMARY))
    }

    #[must_use]
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for ClusterId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

/// Replace the `user:password@` part of every URL inside `text` with `***@`.
/// Cluster errors quote proxy and server URLs; their credentials must never
/// reach logs, health details or terminal output.
#[must_use]
pub fn redact_credentials(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut rest = text;
    while let Some(i) = rest.find("://") {
        let (head, tail) = rest.split_at(i + 3);
        out.push_str(head);
        let end = tail
            .find(|c: char| matches!(c, '/' | '?' | '#' | '"' | '\'' | '<' | '>') || c.is_whitespace())
            .unwrap_or(tail.len());
        let authority = &tail[..end];
        match authority.rfind('@') {
            Some(at) => {
                out.push_str("***@");
                out.push_str(&authority[at + 1..]);
            }
            None => out.push_str(authority),
        }
        rest = &tail[end..];
    }
    out.push_str(rest);
    out
}

/// Mounted into every pod that has a service-account token.
const POD_NAMESPACE_FILE: &str = "/var/run/secrets/kubernetes.io/serviceaccount/namespace";

/// The namespace Kuben itself runs in: `configured` (`kube.namespace`) when
/// set, else the pod's service-account namespace, else `None` (a binary
/// running outside the cluster).
#[must_use]
pub fn own_namespace(configured: Option<&str>) -> Option<String> {
    let non_empty = |s: &str| {
        let s = s.trim();
        (!s.is_empty()).then(|| s.to_owned())
    };
    configured.and_then(non_empty).or_else(|| {
        std::fs::read_to_string(POD_NAMESPACE_FILE)
            .ok()
            .and_then(|s| non_empty(&s))
    })
}

#[derive(Clone)]
pub struct ClusterRegistry {
    clients: Arc<BTreeMap<ClusterId, Client>>,
}

impl fmt::Debug for ClusterRegistry {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("ClusterRegistry")
            .field("clusters", &self.clients.keys().collect::<Vec<_>>())
            .finish()
    }
}

impl ClusterRegistry {
    /// Build the registry for the server. When `kube.required` is false, any
    /// failure (no kubeconfig, unreachable proxy, bad credentials) yields
    /// `Ok(None)` so the API/UI still start degraded for setup and diagnosis.
    pub async fn from_config(cfg: &KubeCfg) -> anyhow::Result<Option<Self>> {
        match Self::connect(cfg).await {
            Ok(registry) => Ok(Some(registry)),
            Err(e) if !cfg.required => {
                tracing::warn!(error = %e, "kubernetes cluster unavailable; starting without cluster");
                Ok(None)
            }
            Err(e) => Err(e),
        }
    }

    /// Connect to the configured cluster. Error messages are redacted.
    pub async fn connect(cfg: &KubeCfg) -> anyhow::Result<Self> {
        Self::try_connect(cfg)
            .await
            .map_err(|e| anyhow::anyhow!(redact_credentials(&format!("{e:#}"))))
    }

    async fn try_connect(cfg: &KubeCfg) -> anyhow::Result<Self> {
        let options = KubeConfigOptions {
            context: cfg.context.clone(),
            ..KubeConfigOptions::default()
        };
        let config = if let Some(path) = &cfg.kubeconfig {
            let kubeconfig = kube::config::Kubeconfig::read_from(path)?;
            Config::from_custom_kubeconfig(kubeconfig, &options).await?
        } else if cfg.context.is_some() {
            Config::from_kubeconfig(&options).await?
        } else {
            Config::infer().await?
        };
        Ok(Self::single(Client::try_from(config)?))
    }

    /// Registry with exactly one cluster.
    #[must_use]
    pub fn single(client: Client) -> Self {
        let mut m = BTreeMap::new();
        m.insert(ClusterId::primary(), client);
        Self { clients: Arc::new(m) }
    }

    #[must_use]
    pub fn get(&self, id: &ClusterId) -> Option<Client> {
        self.clients.get(id).cloned()
    }

    /// The primary cluster (always present).
    ///
    /// # Panics
    /// Never: the registry cannot be constructed without a primary cluster.
    #[must_use]
    pub fn primary(&self) -> Client {
        self.clients
            .get(&ClusterId::primary())
            .cloned()
            .expect("registry always has a primary cluster")
    }

    pub fn ids(&self) -> impl Iterator<Item = &ClusterId> {
        self.clients.keys()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn credentials_are_redacted_from_urls() {
        assert_eq!(
            redact_credentials("configured proxy http://user:s3cret@proxy.local:3128/ requires a feature"),
            "configured proxy http://***@proxy.local:3128/ requires a feature"
        );
        assert_eq!(
            redact_credentials("a postgres://kuben:pw@db:5432/kuben b https://token@api.example.com"),
            "a postgres://***@db:5432/kuben b https://***@api.example.com"
        );
    }

    #[test]
    fn text_without_credentials_is_untouched() {
        for s in [
            "https://10.0.0.1:6443/api",
            "no url here",
            "https://h/path/a@b",
            "trailing ://",
            "",
        ] {
            assert_eq!(redact_credentials(s), s);
        }
    }

    #[test]
    fn configured_namespace_wins() {
        assert_eq!(
            own_namespace(Some(" kuben-system ")).as_deref(),
            Some("kuben-system")
        );
    }

    #[tokio::test]
    async fn unusable_config_degrades_unless_required() {
        let mut cfg = KubeCfg {
            kubeconfig: Some("/nonexistent/kubeconfig".into()),
            ..KubeCfg::default()
        };
        assert!(
            ClusterRegistry::from_config(&cfg)
                .await
                .expect("degrades")
                .is_none()
        );
        cfg.required = true;
        assert!(ClusterRegistry::from_config(&cfg).await.is_err());
    }
}

File: internal/kube/registry/registry.go
// Package registry holds the clients of the clusters Kuben manages
// (crates/kuben-platform/src/registry.rs). It is immutable after
// construction and every operation names its cluster (Invariant I-11);
// there is one cluster, "primary", and the type is keyed so more are
// additive.
package registry

import (
	"fmt"
	"log/slog"
	"os"
	"regexp"
	"strings"

	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"

	"github.com/Teamtem-dev/kuben/internal/core/config"
	"github.com/Teamtem-dev/kuben/internal/core/opt"
)

// Primary is the id of the cluster Kuben runs against.
const Primary = "primary"

// podNamespaceFile is mounted into every pod with a service-account token.
const podNamespaceFile = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"

// Cluster is one cluster's clients.
type Cluster struct {
	ID      string
	Config  *rest.Config
	Typed   kubernetes.Interface
	Dynamic dynamic.Interface
}

// Registry is the clusters by id.
type Registry struct {
	clusters map[string]Cluster
}

// FromConfig builds the registry for the server. Unless kube.required is
// set, a failure (no kubeconfig, bad credentials) yields no registry and
// no error, so the API and the console still start for setup and
// diagnosis.
func FromConfig(cfg config.KubeCfg, logger *slog.Logger) (opt.Val[*Registry], error) {
	r, err := Connect(cfg)
	if err != nil {
		if !cfg.Required {
			logger.Warn("kubernetes cluster unavailable; starting without cluster", "error", err)
			return opt.None[*Registry](), nil
		}
		return opt.None[*Registry](), err
	}
	return opt.Some(r), nil
}

// Connect builds the clients of the configured cluster: kube.kubeconfig
// and kube.context when set, else the usual kubeconfig search ($KUBECONFIG,
// ~/.kube/config), else the in-cluster service account. Error messages are
// redacted.
func Connect(cfg config.KubeCfg) (*Registry, error) {
	rules := clientcmd.NewDefaultClientConfigLoadingRules()
	if path, ok := cfg.Kubeconfig.Get(); ok {
		rules.ExplicitPath = path
	}
	overrides := &clientcmd.ConfigOverrides{}
	if ctx, ok := cfg.Context.Get(); ok {
		overrides.CurrentContext = ctx
	}
	restCfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(rules, overrides).ClientConfig()
	if err != nil {
		return nil, fmt.Errorf("%s", RedactCredentials(err.Error()))
	}
	c, err := NewCluster(Primary, restCfg)
	if err != nil {
		return nil, err
	}
	return Single(c), nil
}

// NewCluster builds the clients of one cluster from its REST config.
func NewCluster(id string, restCfg *rest.Config) (Cluster, error) {
	restCfg = rest.CopyConfig(restCfg)
	restCfg.UserAgent = "kuben"
	typed, err := kubernetes.NewForConfig(restCfg)
	if err != nil {
		return Cluster{}, fmt.Errorf("%s", RedactCredentials(err.Error()))
	}
	dyn, err := dynamic.NewForConfig(restCfg)
	if err != nil {
		return Cluster{}, fmt.Errorf("%s", RedactCredentials(err.Error()))
	}
	return Cluster{ID: id, Config: restCfg, Typed: typed, Dynamic: dyn}, nil
}

// Single is a registry of exactly one cluster, the primary.
func Single(c Cluster) *Registry {
	c.ID = Primary
	return &Registry{clusters: map[string]Cluster{Primary: c}}
}

// Get is the cluster id.
func (r *Registry) Get(id string) (Cluster, bool) {
	c, ok := r.clusters[id]
	return c, ok
}

// Primary is the primary cluster; a registry always has one.
func (r *Registry) Primary() Cluster { return r.clusters[Primary] }

// OwnNamespace is the namespace Kuben runs in: configured (kube.namespace)
// when set, else the pod's service-account namespace, else none (a binary
// outside the cluster).
func OwnNamespace(configured opt.Val[string]) opt.Val[string] {
	if ns := strings.TrimSpace(configured.Or("")); ns != "" {
		return opt.Some(ns)
	}
	data, err := os.ReadFile(podNamespaceFile)
	if err != nil {
		return opt.None[string]()
	}
	if ns := strings.TrimSpace(string(data)); ns != "" {
		return opt.Some(ns)
	}
	return opt.None[string]()
}

var authority = regexp.MustCompile(`://[^/?#"'<>\s]*`)

// RedactCredentials replaces the `user:password@` part of every URL in
// text with `***@`: cluster errors quote proxy and server URLs, whose
// credentials must never reach logs, health details or a terminal.
func RedactCredentials(text string) string {
	return authority.ReplaceAllStringFunc(text, func(m string) string {
		at := strings.LastIndexByte(m, '@')
		if at < 0 {
			return m
		}
		return "://***@" + m[at+1:]
	})
}
